browsergym/core/src/browsergym/core/env.py

```diff
@@ -508,26 +518,16 @@ def _get_obs(self):
         # post-extraction cleanup of temporary info in dom
         _post_extract(self.page)
 
-        # use first user message as goal, if any
-        # use all user images before first user message as goal images, if any
-        goal_msg = "There is no goal."
-        goal_image_urls = []
-        _prev_image_urls = []
-        for msg in self.chat.messages:
-            if msg["role"] == "user_image":
-                _prev_image_urls.append(msg["message"])
-            elif msg["role"] == "user":
-                goal_msg = msg["message"]
-                goal_image_urls = _prev_image_urls
-                break
-            else:
-                pass
+        # if no goal has been set yet, try to extract it from the chat
+        if not self.goal:
+            logger.warning(f"Empty goal, trying to extract goal from chat as a fallback.")
+            self.goal = self._try_and_extract_goal_from_chat()
 
         # obs is generic to all tasks
         obs = {
             "chat_messages": copy.deepcopy(self.chat.messages),
-            "goal": goal_msg,  # TODO: redundant with chat messages, to be removed?
-            "goal_image_urls": goal_image_urls,  # TODO: redundant with chat messages, to be removed?
+            "goal": self._goal_to_text(self.goal),  # legacy goal, deprecated
```

Comment:
minor: make this as a non class function. So it's easier to know inputs and
outputs and also easier to unit test.

Reply:
done